server/service/src/stocktake/insert/generate.rs

```diff
@@ -1,38 +1,103 @@
+) -> Result<(StocktakeRow, Vec<StocktakeLineRow>), RepositoryError> {
+    let stocktake_number = next_number(connection, &NumberRowType::Stocktake, store_id)?;
+    let stocktake_id = input.id.clone();
+    let mut program_id = None;
+
+    let lines = generate_stocktake_lines(
+        connection,
+        store_id,
+        &stocktake_id,
+        InsertStocktakeInput {
+            master_list_id: input.master_list_id.clone(),
+            location: input.location.clone(),
+            items_have_stock: input.items_have_stock,
+            expires_before: input.expires_before,
+            is_initial_stocktake: input.is_initial_stocktake,
+        },
+        &mut program_id,
+    )?;
+
+    let InsertStocktake {
         id,
         comment,
         description,
         stocktake_date,
         is_locked,
-        location,
+        is_initial_stocktake,
+        master_list_id: _,
+        location: _,
+        items_have_stock: _,
+        expires_before: _,
+    } = input;
+
+    Ok((
+        StocktakeRow {
+            id,
+            stocktake_number,
+            comment,
+            description,
+            // TODO: Changing this to be same as created datetime for now since function is disabled in frontend
+            // but will need to remove this later when functionality is
+            stocktake_date: if stocktake_date.is_some() {
+                stocktake_date
+            } else {
+                Some(Utc::now().naive_utc().date())
+            },
+            status: StocktakeStatus::New,
+            created_datetime: Utc::now().naive_utc(),
+            user_id: user_id.to_string(),
+            store_id: store_id.to_string(),
+            is_locked: is_locked.unwrap_or(false),
+            program_id,
+            is_initial_stocktake,
+            // Default
+            finalised_datetime: None,
+            inventory_addition_id: None,
+            inventory_reduction_id: None,
+            counted_by: None,
+            verified_by: None,
+        },
+        lines,
+    ))
+}
+
+fn generate_stocktake_lines(
+    connection: &StorageConnection,
+    store_id: &str,
+    id: &str,
+    InsertStocktakeInput {
         master_list_id,
+        location,
         items_have_stock,
         expires_before,
-    }: InsertStocktake,
-) -> Result<(StocktakeRow, Vec<StocktakeLineRow>), RepositoryError> {
-    let stocktake_number = next_number(connection, &NumberRowType::Stocktake, store_id)?;
-    let mut program_id = None;
-
+        is_initial_stocktake,
+    }: InsertStocktakeInput,
+    program_id: &mut Option<String>,
+) -> Result<Vec<StocktakeLineRow>, RepositoryError> {
+    if is_initial_stocktake {
+        return generate_lines_initial_stocktake(connection, store_id, &id);
+    };
     let master_list_lines = match master_list_id {
         Some(master_list_id) => {
-            program_id = ProgramRowRepository::new(connection)
+            *program_id = ProgramRowRepository::new(connection)
```

Comment:
program_id needs to be saved~ Since you've moved where the stocktake is being generated, you need to move this too~